File: ipaacalib/java/src/ipaaca/Payload.java
/*
 * This file is part of IPAACA, the
 *  "Incremental Processing Architecture
 *   for Artificial Conversational Agents".  
 *
 * Copyright (c) 2009-2013 Sociable Agents Group
 *                         CITEC, Bielefeld University   
 *
 * http://opensource.cit-ec.de/projects/ipaaca/
 * http://purl.org/net/ipaaca
 *
 * This file may be licensed under the terms of of the
 * GNU Lesser General Public License Version 3 (the ``LGPL''),
 * or (at your option) any later version.
 *
 * Software distributed under the License is distributed
 * on an ``AS IS'' basis, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied. See the LGPL for the specific language
 * governing rights and limitations.
 *
 * You should have received a copy of the LGPL along with this
 * program. If not, go to http://www.gnu.org/licenses/lgpl.html
 * or write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  
 *
 * The development of this software was supported by the
 * Excellence Cluster EXC 277 Cognitive Interaction Technology.
 * The Excellence Cluster EXC 277 is a grant of the Deutsche
 * Forschungsgemeinschaft (DFG) in the context of the German
 * Excellence Initiative.
 */

package ipaaca;

import ipaaca.protobuf.Ipaaca.PayloadItem;

import org.apache.commons.lang.StringEscapeUtils;

import com.google.common.collect.ImmutableSet;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Payload of an IU
 * @author hvanwelbergen
 * 
 */
public class Payload implements Map<String, String>
{
    private Map<String, String> map = Collections.synchronizedMap(new HashMap<String, String>());
    private final AbstractIU iu;

    public Payload(AbstractIU iu)
    {
        this.iu = iu;
    }

    public Payload(AbstractIU iu, List<PayloadItem> payloadItems)
    {
        this(iu, payloadItems, null);
    }

    public Payload(AbstractIU iu, Map<String, String> newPayload)
    {
        this(iu, newPayload, null);
    }

    public Payload(AbstractIU iu, Map<String, String> newPayload, String writerName)
    {
        this.iu = iu;
        set(newPayload, writerName);
    }

    public Payload(AbstractIU iu, List<PayloadItem> newPayload, String writerName)
    {
        this.iu = iu;
        set(newPayload, writerName);
    }

    public void set(Map<String, String> newPayload, String writerName)
    {
        iu.setPayload(newPayload, writerName);
        synchronized(map)
        {
            map.clear();
            map.putAll(newPayload);
        }
    }

    public void set(List<PayloadItem> newPayload, String writerName)
    {
        iu.handlePayloadSetting(newPayload, writerName);
        synchronized(map)
        {
            map.clear();
            for (PayloadItem item : newPayload)
            {
                map.put(item.getKey(), pseudoConvertFromJSON(item.getValue(), item.getType()));
            }
        }
    }

    public String pseudoConvertFromJSON(String value, String type) {
        if (type.equals("JSON")) {
            if (value.startsWith("\"")) {
                //return value.replaceAll("\\\"", "");
                return StringEscapeUtils.unescapeJava(value.substring(1, value.length() - 1));
            } else if (value.startsWith("{") || value.startsWith("[") || value.matches("true") || value.matches("false") || value.matches("-?[0-9]*[.,]?[0-9][0-9]*.*")) { 
                return value;
            } else if (value.equals("null")) {
                return "";
            }
        }
        return value;
    }

    void enforcedSetItem(String key, String value)
    {
        map.put(key, value);
    }

    void enforcedRemoveItem(String key)
    {
        map.remove(key);
    }

    public void clear()
    {
        throw new RuntimeException("Not implemented");
        // map.clear();
    }

    public boolean containsKey(Object key)
    {
        return map.containsKey(key);
    }

    public boolean containsValue(Object value)
    {
        return map.containsValue(value);
    }

    /**
     * Provides an immutable copy of the entryset of the Payload
     */
    public ImmutableSet<java.util.Map.Entry<String, String>> entrySet()
    {
        return ImmutableSet.copyOf(map.entrySet());
    }

    public boolean equals(Object o)
    {
        return map.equals(o);
    }

    public String get(Object key)
    {
        return map.get(key);
    }

    public int hashCode()
    {
        return map.hashCode();
    }

    public boolean isEmpty()
    {
        return map.isEmpty();
    }

    public Set<String> keySet()
    {
        return map.keySet();
    }

    /**
     * Set item in this payload.
     * Requests item setting from the OutputBuffer holding the local version
     * of this IU. Returns when permission is granted and item is set;
     * otherwise raises an IUUpdateFailedException.
     */
    public String put(String key, String value, String writer)
    {
        iu.putIntoPayload(key, value, writer);
        return map.put(key, value);
    }

    /**
     * Delete item in this payload.//
     * Requests item deletion from the OutputBuffer holding the local version
     * of this IU. Returns when permission is granted and item is deleted;
     * otherwise raises an IUUpdateFailedError.
     */
    public String remove(Object key, String writer)
    {
        iu.removeFromPayload(key, writer);
        return map.remove(key);
    }

    public String put(String key, String value)
    {
        return put(key, value, null);
    }

    
    public void putAll(Map<? extends String, ? extends String> newItems)
    {
        putAll(newItems, null);
    }

    public void putAll(Map<? extends String, ? extends String> newItems, String writer)
    {
        iu.putIntoPayload(newItems, writer);
   		map.putAll(newItems);
    }
    
    public void merge(Map<? extends String, ? extends String> items) {
    	putAll(items, null);
    }

    public String remove(Object key)
    {
        return remove(key, null);
    }

    public int size()
    {
        return map.size();
    }

    public Collection<String> values()
    {
        return map.values();
    }
    
    @Override
    public String toString()
    {
        return map.toString();
    }
}

File: ipaacalib/python/src/ipaaca/payload.py
# -*- coding: utf-8 -*-

# This file is part of IPAACA, the
#  "Incremental Processing Architecture
#   for Artificial Conversational Agents".
#
# Copyright (c) 2009-2014 Social Cognitive Systems Group
#                         CITEC, Bielefeld University
#
# http://opensource.cit-ec.de/projects/ipaaca/
# http://purl.org/net/ipaaca
#
# This file may be licensed under the terms of of the
# GNU Lesser General Public License Version 3 (the ``LGPL''),
# or (at your option) any later version.
#
# Software distributed under the License is distributed
# on an ``AS IS'' basis, WITHOUT WARRANTY OF ANY KIND, either
# express or implied. See the LGPL for the specific language
# governing rights and limitations.
#
# You should have received a copy of the LGPL along with this
# program. If not, go to http://www.gnu.org/licenses/lgpl.html
# or write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# The development of this software was supported by the
# Excellence Cluster EXC 277 Cognitive Interaction Technology.
# The Excellence Cluster EXC 277 is a grant of the Deutsche
# Forschungsgemeinschaft (DFG) in the context of the German
# Excellence Initiative.

from __future__ import division, print_function

import threading


__all__ = [
	'Payload',
	'PayloadItemDictProxy',
	'PayloadItemListProxy',
]


_DEFAULT_PAYLOAD_UPDATE_TIMEOUT = 0.1


class Payload(dict):

	def __init__(self, iu, writer_name=None, new_payload=None, omit_init_update_message=False, update_timeout=_DEFAULT_PAYLOAD_UPDATE_TIMEOUT):
		self.iu = iu
		_pl = {}
		for k, v in ({} if new_payload is None else new_payload).iteritems():
			_pl[unicode(k, 'utf8') if type(k) == str else k] = unicode(v, 'utf8') if type(v) == str else v
		# NOTE omit_init_update_message is necessary to prevent checking for
		#   exceptions and sending updates in the case where we just receive
		#   a whole new payload from the remote side and overwrite it locally.
		for k, v in _pl.iteritems():
			dict.__setitem__(self, k, v)
		if (not omit_init_update_message) and (self.iu.buffer is not None):
			self.iu._modify_payload(
				is_delta=False,
				new_items=_pl,
				keys_to_remove=[],
				writer_name=writer_name)
		self._update_on_every_change = True
		self._update_timeout = update_timeout
		self._collected_modifications = {}
		self._collected_removals = []
		self._batch_update_writer_name = None  # name of remote buffer or None
		self._batch_update_lock = threading.RLock()
		self._batch_update_cond = threading.Condition(threading.RLock())

	def __getitem__(self, k):
		value = dict.__getitem__(self, k)
		if isinstance(value, dict):
			return PayloadItemDictProxy(value, self, k)
		elif isinstance(value, list):
			return PayloadItemListProxy(value, self, k)
		else:
			return value

	def __setitem__(self, k, v, writer_name=None):
		with self._batch_update_lock.acquire:
			for k, v in payload.iteritems():
				k = unicode(k, 'utf8') if type(k) == str else k
				v = unicode(v, 'utf8') if type(v) == str else v
			if self._update_on_every_change:
				self.iu._modify_payload(
					is_delta=True,
					new_items={k:v},
					keys_to_remove=[],
					writer_name=writer_name)
			else: # Collect additions/modifications
				self._batch_update_writer_name = writer_name
				self._collected_modifications[k] = v
			return dict.__setitem__(self, k, v)

	def __delitem__(self, k, writer_name=None):
		with self._batch_update_lock.acquire:
			k = unicode(k, 'utf8') if type(k) == str else k
			if self._update_on_every_change:
				self.iu._modify_payload(
					is_delta=True,
					new_items={},
					keys_to_remove=[k],
					writer_name=writer_name)
			else: # Collect additions/modifications
				self._batch_update_writer_name = writer_name
				self._collected_removals.append(k)
			return dict.__delitem__(self, k)

	# Context-manager based batch updates, not thread-safe (on remote updates)
	def __enter__(self):
		self._wait_batch_update_lock(self._update_timeout)
		self._update_on_every_change = False

	# Context-manager based batch updates, not thread-safe (on remote updates)
	def __exit__(self, type, value, traceback):
		self.iu._modify_payload(
			is_delta=True,
			new_items=self._collected_modifications,
			keys_to_remove=self._collected_removals,
			writer_name=self._batch_update_writer_name)
		self._collected_modifications = {}
		self._collected_removals = []
		self._update_on_every_change = True
		self._batch_update_writer_name = None
		self._batch_update_lock.release()

	def merge(self, payload, writer_name=None):
		with self._batch_update_lock.acquire:
			for k, v in payload.iteritems():
				k = unicode(k, 'utf8') if type(k) == str else k
				v = unicode(v, 'utf8') if type(v) == str else v
			self.iu._modify_payload(
				is_delta=True,
				new_items=payload,
				keys_to_remove=[],
				writer_name=writer_name)
			return dict.update(self, payload) # batch update

	def _remotely_enforced_setitem(self, k, v):
		"""Sets an item when requested remotely."""
		return dict.__setitem__(self, k, v)

	def _remotely_enforced_delitem(self, k):
		"""Deletes an item when requested remotely."""
		return dict.__delitem__(self, k)

	def _wait_batch_update_lock(self, timeout):
		# wait lock with time-out http://stackoverflow.com/a/8393033
		with self._batch_update_cond:
			current_time = start_time = time.time()
			while current_time < start_time + timeout:
				if self._batch_update_lock.acquire(False):
					return True
				else:
					self._batch_update_cond.wait(timeout - current_time + start_time)
					current_time = time.time()
		raise IUPayloadLockTimeoutError(self.iu)


class PayloadItemProxy(object):

	def __init__(self, content, payload, identifier_in_payload):
		self.payload = payload
		self.content = content
		self.identifier_in_payload = identifier_in_payload

	def _notify_payload(self):
		self.payload[self.identifier_in_payload] = self.content

	def _create_proxy(self, obj, identifier_in_payload):
		if isinstance(obj, dict):
			return PayloadItemDictProxy(obj, self.payload, identifier_in_payload)
		elif isinstance(obj, list):
			return PayloadItemListProxy(obj, self.payload, identifier_in_payload)
		else:
			return obj

	def __setitem__(self, k, v):
		self.content.__setitem__(k,v)
		self._notify_payload()

	def __getitem__(self, k):
		item = self.content.__getitem__(k)
		return self._create_proxy(item, k)

	def __delitem__(self, k):
		self.content.__delitem__(k)
		self._notify_payload()


class PayloadItemDictProxy(PayloadItemProxy, dict):

	def __init__(self, content, payload, identifier_in_payload):
		dict.__init__(self, content)
		PayloadItemProxy.__init__(self, content, payload, identifier_in_payload)

	def clear(self):
		self.content.clear()
		self._notify_payload()

	def get(self, key, default=None):
		value = self.content.get(key, default)
		return self._create_proxy(value, key)

	def items(self):
		return [(key, value) for key, value in self.iteritems()]

	def iteritems(self):
		for key, value in self.content.iteritems():
			yield key, self._create_proxy(value, key)

	def values(self):
		return [value for value in self.itervalues()]

	def itervalues(self):
		for key, value in self.content.iteritems():
			yield self._create_proxy(value, key)

	def pop(self, key, *args):
		x = self.content.pop(key, *args)
		self._notify_payload()
		return x

	def popitem(self):
		x = self.content.popitem()
		self._notify_payload()
		return x

	def setdefault(self, key, default=None):
		notification_necessary = not key in self.content
		x = self.content.setdefault(key, default)
		if notification_necessary:
			self._notify_payload()
		return x

	def update(self, *args, **kwargs):
		self.content.update(*args, **kwargs)
		self._notify_payload()


class PayloadItemListProxy(PayloadItemProxy, list):

	def __init__(self, content, payload, identifier_in_payload):
		list.__init__(self, content)
		PayloadItemProxy.__init__(self, content, payload, identifier_in_payload)

	def __iter__(self):
		for index, item in enumerate(self.content):
			yield self._create_proxy(item, index)

	def append(self, x):
		self.content.append(x)
		self._notify_payload()

	def extend(self, l):
		self.content.extend(l)
		self._notify_payload()

	def insert(self, i, x):
		self.content.insert(i, x)
		self._notify_payload()

	def remove(self, x):
		self.content.remove(x)
		self._notify_payload()

	def pop(self, *args, **kwargs):
		x = self.content.pop(*args, **kwargs)
		self._notify_payload()
		return x

	def sort(self, cmp=None, key=None, reverse=False):
		self.content.sort(cmp, key, reverse)
		self._notify_payload()

	def reverse(self):
		self.content.reverse()
		self._notify_payload()
